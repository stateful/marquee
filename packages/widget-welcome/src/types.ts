import { ContextProperties } from "@vscode-marquee/utils";

export interface Trick {
  order: number;
  id: string;
  content: string;
  title: string;
  active: boolean;
  notify: boolean;
  createdAt: number;
  votes: { upvote: number }
}

export interface Context extends ContextProperties<State> {
  _setLiked: (id: string) => void
  _setRead: (id: string) => void
  _resetRead: () => void
};

export interface State {
  read: string[];
  liked: string[];
  tricks: Trick[];
  error: Error | null;
}

export interface Configuration {}

export interface Events {
  upvote: string;
}

export interface Storage extends Pick<Context, 'read' | 'liked'> {}
