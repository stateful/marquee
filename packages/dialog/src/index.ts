import DialogContainer from './DialogContainer'
import DialogTitle from './DialogTitle'

export { DialogContainer, DialogTitle }
