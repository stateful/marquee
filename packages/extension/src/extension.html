<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta
      http-equiv="Content-Security-Policy"
      content="
        default-src 'unsafe-inline';
        img-src 'self' data: <%= it.cspSource %> https:;
        font-src <%= it.cspSource %> https:;
        script-src 'nonce-<%= it.nonce %>';
        connect-src <%= it.connectSrc.join(' ') %> 'unsafe-eval';
        child-src <%= it.childSrc.join(' ') %>;
      "
    >
    <base href="<%= it.baseAppUri %>">
    <title>Marquee</title>
    <style>
        .MuiTypography-h1,
        .MuiTypography-h2,
        .MuiTypography-h3,
        .MuiTypography-h4,
        .MuiTypography-h5,
        .MuiTypography-h6,
        .MuiTypography-subtitle1,
        .MuiTypography-subtitle2,
        .MuiTypography-body1,
        .MuiTypography-body2,
        .MuiTypography-button,
        .MuiTypography-caption,
        .MuiTypography-overline {
            font-size: <%= it.fontSize %>em!important;
        }
    </style>
    <style>
      :root {
        --marquee-theme-color: <%= it.colorScheme %>;
      }
    </style>
    <script nonce="<%= it.nonce %>">
      function b64DecodeUnicode(str) {
        // Going backwards: from bytestream, to percent-encoding, to original string.
        return decodeURIComponent(atob(str).split('').map(function(c) {
            return '%' + ('00' + c.charCodeAt(0).toString(16)).slice(-2);
        }).join(''));
      }
      window.marqueeThirdPartyWidgets = <%= it.customWidgetCounter %>;
      window.activeWorkspace = <%~ JSON.stringify(it.aws) %>;
      window.marqueeUserProps = '<%~ JSON.stringify(it.props) %>';
      window.marqueeBackendBaseUrl = '<%= it.baseUrl %>';
      window.marqueeBackendGeoUrl = '<%= it.geoUrl %>';
      window.marqueeBackendFwdGeoUrl = '<%= it.fwdGeoUrl %>';
      window.marqueeStateConfiguration = JSON.parse(b64DecodeUnicode('<%= it.widgetStateConfigurations %>'));
    </script>
</head>
<body>
    <div id="root"></div>
    <script nonce="<%= it.nonce %>" src="<%= it.baseAppUri %>/main.js"></script>
    <% for (const widgetScript of it.widgetScripts) { %>
      <%~ widgetScript %>
    <% } %>
</body>
</html>
