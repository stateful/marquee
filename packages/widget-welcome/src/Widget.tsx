import React, { useContext } from 'react'
import { Grid, Link, Typography } from '@mui/material'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faDiscord } from '@fortawesome/free-brands-svg-icons/faDiscord'

import wrapper, { Dragger, HeaderWrapper } from '@vscode-marquee/widget'
import { NetworkError } from '@vscode-marquee/utils'
import type { MarqueeWidgetProps } from '@vscode-marquee/widget'

import TrickContext, { TrickProvider } from './Context'
import TrickContent from './components/TrickContent'
import PopMenu from './components/Pop'

const WidgetBody = () => {
  const { error } = useContext(TrickContext)
  const link = 'https://discord.gg/BQm8zRCBUY'
  return (
    <Grid item xs>
      <Grid
        container
        wrap="nowrap"
        direction="column"
        style={{ height: '100%' }}
      >
        {error && (
          <Grid
            item
            xs
            style={{
              overflow: 'auto',
              height: '100%',
              width: '100%',
              padding: '24px',
            }}
          >
            <NetworkError message={error.message} />
          </Grid>
        )}
        {!error && (
          <Grid
            item
            xs
            style={{
              overflow: 'auto',
              paddingTop: '16px',
              paddingRight: '16px',
              paddingLeft: '16px',
              paddingBottom: '8px',
            }}
          >
            <Grid
              container
              style={{
                height: '100%',
                width: '100%',
              }}
              direction="column"
              wrap="nowrap"
            >
              <Grid item xs={1} style={{ maxWidth: '100%' }}>
                <Grid
                  container
                  justifyContent="flex-end"
                  alignItems="center"
                  style={{ height: '100%', padding: '8px' }}
                >
                  <Grid item>
                    <Typography variant="caption">
                      <Link
                        component="a"
                        href={link}
                        target="_blank"
                        underline="hover">
                        Join Discord &nbsp;
                        <FontAwesomeIcon icon={faDiscord} />
                      </Link>
                    </Typography>
                  </Grid>
                </Grid>
              </Grid>
              <Grid item xs={11} style={{ maxWidth: '100%' }}>
                <TrickContent />
              </Grid>
            </Grid>
          </Grid>
        )}
      </Grid>
    </Grid>
  )
}

let Welcome = ({ ToggleFullScreen }: MarqueeWidgetProps) => {
  return (
    <>
      <HeaderWrapper>
        <Grid item>
          <Typography variant="subtitle1">Mailbox</Typography>
        </Grid>
        <Grid item>
          <Grid container direction="row" spacing={1}>
            <Grid item>
              <PopMenu />
            </Grid>
            <Grid item>
              <ToggleFullScreen />
            </Grid>
            <Grid item>
              <Dragger />
            </Grid>
          </Grid>
        </Grid>
      </HeaderWrapper>
      <WidgetBody />
    </>
  )
}

const Widget = (props: any) => (
  <TrickProvider>
    <Welcome {...props} />
  </TrickProvider>
)
export default wrapper(Widget, 'welcome')
