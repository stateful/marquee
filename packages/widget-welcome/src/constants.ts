import type { State, Events } from './types'

export const DEFAULT_STATE: State & Omit<Events, 'upvote'> = {
  tricks: [],
  read: [],
  liked: [],
  error: null
}
