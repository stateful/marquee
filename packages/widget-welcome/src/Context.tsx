import React, { createContext, useEffect } from 'react'
import { connect, getEventListener, MarqueeWindow } from '@vscode-marquee/utils'

import { DEFAULT_STATE } from './constants'
import type { Context, State, Events } from './types'

declare const window: MarqueeWindow

interface Props {
  children?: React.ReactNode;
}

const TrickContext = createContext<Context>(DEFAULT_STATE as any as Context)
const WIDGET_ID = '@vscode-marquee/welcome-widget'

const TrickProvider = ({ children }: Props) => {
  const widgetState = getEventListener<State>(WIDGET_ID)
  const widgetEvents = getEventListener<Events>(WIDGET_ID)
  const providerValues = connect<State>(window.marqueeStateConfiguration[WIDGET_ID].state, widgetState)

  const _setRead = (id: string) => {
    if (!providerValues.read.includes(id)) {
      providerValues.setRead([...providerValues.read, id])
    }
  }

  const _setLiked = (id: string) => {
    if (!providerValues.liked.includes(id)) {
      providerValues.setLiked([...providerValues.liked, id])
      widgetEvents.emit('upvote', id)
    }
  }

  const _resetRead = () => {
    providerValues.setRead([])
  }

  useEffect(() => {
    return () => {
      widgetState.removeAllListeners()
      widgetEvents.removeAllListeners()
    }
  }, [])

  return (
    <TrickContext.Provider
      value={{
        ...providerValues,
        _setLiked,
        _setRead,
        _resetRead,
      }}
    >
      {children}
    </TrickContext.Provider>
  )
}

export default TrickContext
export { TrickProvider }
