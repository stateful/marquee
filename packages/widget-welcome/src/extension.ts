import vscode from 'vscode'
import Axios, { AxiosRequestConfig } from 'axios'
import axiosRetry from 'axios-retry'
import ExtensionManager, { Logger, ChildLogger } from '@vscode-marquee/utils/extension'
import { Client } from 'tangle'

import { DEFAULT_STATE } from './constants'
import type { State, Events, Configuration, Trick } from './types'

declare const BACKEND_BASE_URL: string

const STATE_KEY = 'widgets.welcome'
const FETCH_INTERVAL = 5 * 1000 * 60 // 5min
const AXIOS_RETRIES = process.env.NODE_ENV === 'development'
  ? 1
  : 10
const config = vscode.workspace.getConfiguration('marquee')
axiosRetry(Axios, { retries: AXIOS_RETRIES, retryDelay: axiosRetry.exponentialDelay })

class StateManager extends ExtensionManager<State & Events, Configuration> {
  #logger: ChildLogger

  private _interval: NodeJS.Timeout
  private _prevtricks?: Trick[]
  private _retainTricks: boolean = false

  constructor (context: vscode.ExtensionContext) {
    super(context, STATE_KEY, {}, DEFAULT_STATE as State & Events)
    this.#logger = Logger.getChildLogger(STATE_KEY)
    this.fetchData()
    this._interval = setInterval(this.fetchData.bind(this), FETCH_INTERVAL)
  }

  get backendUrl () {
    return process.env.NODE_ENV === 'test' ? 'http://test' : BACKEND_BASE_URL
  }

  /**
   * broadcast to webview if available
   * @param state to broadcast
   */
  broadcast (state: Partial<State & Events>) {
    this.emit('state', state)

    if (!this._tangle) {
      return
    }

    this._tangle.broadcast(state as State & Events)
  }

  /**
   * Make sure we have set proper proxy settings
   * @returns AxiosRequestConfig
   */
  private _getRequestOptions () {
    const pref: any = config.get('configuration')
    const options: AxiosRequestConfig = {}

    if (pref?.proxy) {
      const p = new URL(pref.proxy)
      options.proxy = {
        protocol: p.protocol,
        host: p.hostname,
        port: parseInt(p.port),
        auth: {
          username: p.username,
          password: p.password
        }
      }
    }

    return options
  }

  /**
   * fetch data and broadcast them across the Marquee app
   */
  async fetchData () {
    const url = `${this.backendUrl}/getTricks`
    this.#logger.info(`Fetching ${url}`)
    const result = await Axios.get(url, this._getRequestOptions()).then(
      (res) => res.data as Trick[],
      (err) => err as Error
    )

    if (result instanceof Error) {
      this.#logger.error(`Error fetching tricks: ${result.message}`)
      return this.broadcast({ error: result })
    }

    if (this._prevtricks && this._prevtricks.length < result.length) {
      const newTrick = result.slice(this._prevtricks.length)
        .filter((trick) => trick.notify && trick.active)
        .pop()

      if (newTrick) {
        this.#logger.info(`Notify new trick: ${newTrick.title}`)
        vscode.window
          .showInformationMessage(newTrick.title, 'Learn more')
          .then(() => this.emit('gui.open'))
      }
    }

    /**
     * only broadcast if we haven't before or a new trick was received
     */
    const prevTrickIds = new Set(this._prevtricks?.map((trick) => trick.id) ?? [])
    const netNewTrickIds = new Set(result.filter(resTrick => !prevTrickIds.has(resTrick.id)))
    if (this._retainTricks && (!this._prevtricks || netNewTrickIds.size > 0)) {
      this._prevtricks = result
      this.#logger.info(`Broadcast ${result.length} tricks`)
      this.broadcast({ error: null, tricks: result })
    }
  }

  private _upvoteTrick (id: string) {
    this.#logger.info(`Upvote trick with id: ${id}`)
    return Axios.post(
      `${this.backendUrl}/voteTrick`,
      { op: 'upvote', id },
      this._getRequestOptions()
    ).catch((err) => vscode.window.showErrorMessage('Failed to upvote trick!', err.message))
  }

  setBroadcaster (tangle: Client<State & Events>) {
    super.setBroadcaster(tangle)
    tangle.whenReady().then(() => {
      this._retainTricks = true
      return this.fetchData()
    })
    tangle.on('upvote', this._upvoteTrick.bind(this))
    return this
  }

  dispose () {
    super.dispose()
    clearInterval(this._interval)
  }
}

let stateManager: StateManager
export function activate (context: vscode.ExtensionContext) {
  stateManager = new StateManager(context)

  // don't allow errors to be recovered from default state
  const defaultState = stateManager.state
  if (defaultState.error) {
    defaultState.error = null
  }

  return {
    marquee: {
      disposable: stateManager,
      defaultState,
      defaultConfiguration: stateManager.configuration,
      setup: stateManager.setBroadcaster.bind(stateManager)
    }
  }
}
