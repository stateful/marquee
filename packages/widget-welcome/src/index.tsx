import React from 'react';
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faMailBulk } from "@fortawesome/free-solid-svg-icons/faMailBulk";

import NavPop from "./components/NavPop";
import Welcome from "./Widget";
import { TrickProvider } from "./Context";

export default {
  name: "welcome",
  icon: <FontAwesomeIcon icon={faMailBulk} />,
  label: "Mailbox",
  tags: ["news", "product", "communication", "tips", "tricks"],
  description: "Where to look for Marquee news, tips and tricks.",
  component: Welcome,
};
export { Welcome, NavPop, TrickProvider };
