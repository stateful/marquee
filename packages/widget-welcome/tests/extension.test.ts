import Axios from 'axios'
import vscode from 'vscode'
import { activate } from '../src/extension'

jest.mock('axios-retry', () => jest.fn().mockReturnValue(Promise.resolve(Axios)))

jest.mock('axios', () => ({
  get: jest.fn().mockResolvedValue({ data: [] }),
  post: jest.fn().mockResolvedValue({ data: [] })
}))

jest.mock('vscode', () => ({
  ConfigurationTarget: { Global: 1 },
  workspace: {
    getConfiguration: jest.fn().mockReturnValue(new Map()),
    onDidChangeConfiguration: jest.fn()
  },
  window: {
    createOutputChannel: jest.fn().mockReturnValue({ appendLine: jest.fn() }),
    showInformationMessage: jest.fn().mockResolvedValue({}),
    showErrorMessage: jest.fn().mockResolvedValue({})
  }
}))
const channel = { appendLine: jest.fn() }

test('should return expected interface', async () => {
  const tangle: any = {
    on: jest.fn(),
    emit: jest.fn(),
    broadcast: jest.fn(),
    listen: jest.fn().mockReturnValue({ unsubscribe: jest.fn() }),
    whenReady: jest.fn().mockReturnValue(Promise.resolve()),
    removeAllListeners: jest.fn()
  }
  const context = { globalState: new Map() }
  // @ts-expect-error
  context.globalState.setKeysForSync = jest.fn()
  context.globalState.set('persistence', {})
  const result = activate(context as any, channel as any)

  expect(result.marquee).toBeTruthy()
  result.marquee.setup(tangle)

  expect(Axios.get).toBeCalledWith('http://test/getTricks', {})
  expect(tangle.on).toBeCalledWith('upvote', expect.any(Function))

  const upvoteTrick = tangle.on.mock.calls.pop().pop()
  const config = vscode.workspace.getConfiguration()
  config.set('configuration', { proxy: 'http://someproxy:8080' })
  upvoteTrick(100)
  expect(Axios.post).toBeCalledWith('http://test/voteTrick', {
    id: 100,
    op: 'upvote'
  }, {
    proxy: {
      protocol: 'http:',
      host: 'someproxy',
      port: 8080,
      auth: { username: '', password: '' }
    }
  });

  (Axios.post as jest.Mock).mockRejectedValue(new Error('ups'))
  upvoteTrick(100)
  await new Promise((resolve) => setTimeout(resolve, 100))
  expect(vscode.window.showErrorMessage)
    .toBeCalledWith('Failed to upvote trick!', 'ups')

  result.marquee.disposable.emit = jest.fn();
  (Axios.get as jest.Mock).mockResolvedValue({
    data: [{
      notify: true,
      active: true,
      title: 'foobar'
    }]
  })
  result.marquee.disposable.fetchData()
  await new Promise((resolve) => setTimeout(resolve, 100))
  expect(vscode.window.showInformationMessage).toBeCalledWith('foobar', 'Learn more')
  expect(result.marquee.disposable.emit).toBeCalledWith('gui.open');

  (Axios.get as jest.Mock).mockRejectedValue(new Error('ups'))
  result.marquee.disposable.fetchData()
  await new Promise((resolve) => setTimeout(resolve, 100))
  expect((result.marquee.disposable.broadcast as jest.Mock).mock.calls)
    .toMatchSnapshot()
})
