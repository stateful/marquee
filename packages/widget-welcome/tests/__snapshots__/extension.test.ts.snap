// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return expected interface 1`] = `
Array [
  Array [
    Object {
      "error": null,
      "tricks": Array [],
    },
  ],
  Array [
    Object {
      "error": null,
      "tricks": Array [
        Object {
          "active": true,
          "notify": true,
          "title": "foobar",
        },
      ],
    },
  ],
  Array [
    Object {
      "error": [Error: ups],
    },
  ],
]
`;
